BUGFIX: `WorkspaceService` adjust to real world use cases

mhsdesign opened the pull request:
The adjustments which are needed for the new workspace UI https://github.com/neos/neos-development-collection/pull/5132

Fix up for https://github.com/neos/neos-development-collection/pull/5146 (original introduction of mighty `WorkspaceService`), and https://github.com/neos/neos-development-collection/pull/5298 (partial refactoring yet again towards authentication provider)


As separate PR to have the bugfixes tested and also better discussed.

Todo:
- [x] write test for workspace delete
- [ ] test for workspace low level pruning? -> should be tested as part of site:pruneAll
- [x] test for live publisher (via "normal" editor)
- [x] test to change base workspace
- [x] cleanup old policy yaml, and explain "special" magic features (like live publisher is added as role on creation for live)
- [x] Discuss deletion should also work from core command? Maybe catchup hook or delete dangling workspace metadata on creation of new identical

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

mhsdesign committed:
TASK: Introduce utility methods for #5132

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -83,6 +84,12 @@ public function getWorkspaceMetadata(ContentRepositoryId $contentRepositoryId, W
      */
     public function setWorkspaceTitle(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, WorkspaceTitle $newWorkspaceTitle): void
     {
+        $workspaceMetadata = $this->loadWorkspaceMetadata($contentRepositoryId, $workspaceName);
```
This can return `null`. I can fix this tomorrow

mhsdesign replied:
i think that fetching it here is just needless no? can we not just put the upsert into one query below? And not rely on the affected rows?

I thought it would just be simple to fix :D now im stuck reading through things like this https://sqlperformance.com/2020/09/locking/upsert-anti-pattern 😂

mhsdesign replied:
and a test like 

```
Scenario: Change title of a root workspace to the same again
  When the root workspace "some-root-workspace" is created
  And the title of workspace "some-root-workspace" is set to "Some new workspace title"
  Then the workspace "some-root-workspace" should have the following metadata:
    | Title                    | Description | Classification | Owner user id |
    | Some new workspace title |             | ROOT           |               |
  And the title of workspace "some-root-workspace" is set to "Some new workspace title"
  Then the workspace "some-root-workspace" should have the following metadata:
    | Title                    | Description | Classification | Owner user id |
    | Some new workspace title |             | ROOT           |               |
```

that it does NOT fail ... maybe we can catch the unique constraint error :D

mhsdesign replied:
Fixed via https://github.com/neos/neos-development-collection/pull/5397

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -83,6 +84,12 @@ public function getWorkspaceMetadata(ContentRepositoryId $contentRepositoryId, W
     public function setWorkspaceTitle(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, WorkspaceTitle $newWorkspaceTitle): void
     {
+        $workspaceMetadata = $this->loadWorkspaceMetadata($contentRepositoryId, $workspaceName);
+        if ($workspaceMetadata->title->equals($newWorkspaceTitle)) {
```
This would be the fix I guess!?
```suggestion
        if ($workspaceMetadata?->title->equals($newWorkspaceTitle)) {
```

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -93,6 +100,12 @@ public function setWorkspaceTitle(ContentRepositoryId $contentRepositoryId, Work
     public function setWorkspaceDescription(ContentRepositoryId $contentRepositoryId, WorkspaceName $workspaceName, WorkspaceDescription $newWorkspaceDescription): void
     {
+        $workspaceMetadata = $this->loadWorkspaceMetadata($contentRepositoryId, $workspaceName);
+        if ($workspaceMetadata->description->equals($newWorkspaceDescription)) {
```
```suggestion
        if ($workspaceMetadata?->description->equals($newWorkspaceDescription)) {
```

mhsdesign committed:
TASK: Introduce `WorkspaceService::deleteWorkspace`
to also removed assigned metadata and roles

mhsdesign commented:
Also im not sure why these changes are needed but they were made originally in 916e51009e9c97b42aa32372b94d149a2022faeb:

```patch
diff --git a/Neos.Neos/Classes/Domain/Service/WorkspaceService.php b/Neos.Neos/Classes/Domain/Service/WorkspaceService.php
index 8d7e11d2e4f..d515a0205cc 100644
--- a/Neos.Neos/Classes/Domain/Service/WorkspaceService.php
+++ b/Neos.Neos/Classes/Domain/Service/WorkspaceService.php
@@ -302,19 +302,30 @@ public function getWorkspacePermissionsForUser(ContentRepositoryId $contentRepos
         } catch (NoSuchRoleException $e) {
             throw new \RuntimeException(sprintf('Failed to determine roles for user "%s", check your package dependencies: %s', $user->getId()->value, $e->getMessage()), 1727084881, $e);
         }
+
+        $userIsAdministrator = in_array('Neos.Neos:Administrator', $userRoles, true);
+
+        if ($userIsAdministrator) {
+            return WorkspacePermissions::all();
+        }
+
         $workspaceMetadata = $this->loadWorkspaceMetadata($contentRepositoryId, $workspaceName);
-        if ($workspaceMetadata !== null && $workspaceMetadata->ownerUserId !== null && $workspaceMetadata->ownerUserId->equals($user->getId())) {
+        $userIsOwner = $workspaceMetadata !== null && $workspaceMetadata->ownerUserId !== null && $workspaceMetadata->ownerUserId->equals($user->getId());
+
+        if ($userIsOwner) {
             return WorkspacePermissions::all();
         }
+
         $userWorkspaceRole = $this->loadWorkspaceRoleOfUser($contentRepositoryId, $workspaceName, $user->getId(), $userRoles);
-        $userIsAdministrator = in_array('Neos.Neos:Administrator', $userRoles, true);
+
         if ($userWorkspaceRole === null) {
-            return WorkspacePermissions::create(false, false, $userIsAdministrator);
+            return WorkspacePermissions::none();
         }
+
         return WorkspacePermissions::create(
             read: $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
             write: $userWorkspaceRole->isAtLeast(WorkspaceRole::COLLABORATOR),
-            manage: $userIsAdministrator || $userWorkspaceRole->isAtLeast(WorkspaceRole::MANAGER),
+            manage: $userWorkspaceRole->isAtLeast(WorkspaceRole::MANAGER),
         );
     }
```

a lot of test fail with this change and im not sure what the reason for this was.

mhsdesign committed:
TASK: Test `ChangeBaseWorkspace` behaviour with security

mhsdesign committed:
TASK: Ensure the live workspace is created by the WorkspaceService for permissions test

mhsdesign committed:
TASK: Ensure the live workspace is created by the WorkspaceService for permissions test

mhsdesign committed:
TASK: Allow workspace creation with only base workspace read privilege

as change base workspace would allow either way to repoint the base

mhsdesign committed:
BUGFIX: Only allow to publish if the current user can write to the base

mhsdesign committed:
TASK: Simplify test to only have one `uninvolved_editor`

mhsdesign committed:
BUGFIX: Ensure that the live workspace is created WITH public viewing permission

mhsdesign commented:
Also it seems, that if we allow to create workspaces only with read permission this test does not pass because EVERY Neos user is EVERYBODY and they are VIEWER

```feature
Scenario Outline: Creating a workspace without READ permissions (on live)
  Given I am authenticated as <user>
  And the shared workspace "some-shared-workspace" is created with the target workspace "live"
  Then an exception of type "AccessDenied" should be thrown with code 1729086686

  And the personal workspace "some-other-personal-workspace" is created with the target workspace "live" for user <user>
  Then an exception of type "AccessDenied" should be thrown with code 1729086686

  Examples:
    | user              |
    | restricted_editor |
    | simple_user       |
```

Technically - if would test this which we dont - this meand that even an unauthenticated request could just create workspaces? - we definitely need test for the not authenticated case as well imo, especially since the `default => grant` would be dangerous no? -> https://github.com/neos/neos-development-collection/blob/484c1abfd1602f70bd95b10988c3ceb3cce9f41c/Neos.Neos/Classes/Security/ContentRepositoryAuthProvider/ContentRepositoryAuthProvider.php#L173

mhsdesign committed:
TASK: Add test for not authenticated case

(https://github.com/neos/neos-development-collection/pull/5334#issuecomment-2523611432)

mhsdesign committed:
TASK: Fix workspace permission test to no be in authorizationChecksDisabled mode after enabling security

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/ContentRepository/Security/WorkspacePermissions.feature:
```diff
@@ -112,15 +100,62 @@ Feature: Workspace permission related features
       | collaborator |
       | owner        |
 
+  Scenario: Creating a workspace without Neos User but READ permissions on live
+    Given I am not authenticated
+    And the shared workspace "some-shared-workspace" is created with the target workspace "live"
+
+  Scenario Outline: Creating a workspace with READ permissions (on live)
+    Given I am authenticated as <user>
+    And the shared workspace "some-shared-workspace" is created with the target workspace "live"
+
+    And the personal workspace "some-other-personal-workspace" is created with the target workspace "live" for user <user>
+
+    Examples:
+      | user              |
+      | admin             |
+      | owner             |
+      | collaborator      |
+      | uninvolved_editor |
+      | restricted_editor |
+      | simple_user       |
```
basically theses test means that EVERYBODY can create a new workspace based on live, not even a neos user is required - im not sure if this is correct, but in 8.3 i also could find no restriction for `new Workspace` or adding that to the repository.

bwaidelich commented:
@mhsdesign thanks for extracting this PR!

> im not sure why these changes are needed

They make me suspicious, too. IMO an admin should be able to grant themselves permissions to write to other personal workspaces, but they should not have this permission _automagically_.

> we definitely need test for the not authenticated case as well imo [...]

Yes, absolutely. I'm not sure, who is _supposed_ to be able to create workspaces – maybe we need an explicit role/privilege for that

> [...] especially since the default => grant would be dangerous no?

The default of grant just matches commands that don't exist yet (because other cases should be covered). That _is_ potentially dangerous ofc (I tried to create an "allowlist"-test but it's not easily possible to get all instances of the `CommandInterface` without our Flow ReflectionService). But AFAIS that is not related to the unauthenticated case

mhsdesign commented:
regarding that not authenticated users can create workspaces based on live (see https://github.com/neos/neos-development-collection/pull/5334/commits/28b98c50c69238e659f16cb6827299b5ee9cd782)

maybe that is okay as in 8.3 that was only restricted via the endpoint and `Neos.Neos:Backend.CreateWorkspaces` now in 9.0 renamed to `Neos.Workspace.Ui:Backend.CreateWorkspaces`:

https://github.com/neos/neos-development-collection/blob/67c9ec4d2c1a198786b18a3da194edeca2bce5da/Neos.Workspace.Ui/Configuration/Policy.yaml#L9C6-L11

So the privilege still exists and covers the default endpoint and usecase. Still having a custom controller _could_ contain code creating a workspace which is then not restricted by the cr. As long as this is the same behaviour as in 8.3 that might be just valid.

---------------

> But AFAIS that is not related to the unauthenticated case

it is imo, because then we never ensure that the roles contain `Authenticated` or that a user exists...

mhsdesign committed:
TASK: Use `deleteWorkspace` from service in cli command and test

mhsdesign committed:
TASK: Throw specific `WorkspaceDoesNotExist` in workspace service to simplify handling

mhsdesign committed:
BUGFIX: Allow to specify `WorkspaceRoleAssignments` when creating a new shared workspace

... otherwise authentication fails and the workspace roles cannot be assigned because management is not granted (yet) - only admins would be able to do that.

mhsdesign commented:
Fixed a conceptional bug to allow to specify `WorkspaceRoleAssignments` when creating a new shared workspace

otherwise authorisation fails when trying to issue `assignWorkspaceRole`: the workspace roles cannot be assigned because management is not granted (yet) - only admins would be able to do that.

I fixed that nicely by allowing to specify the `WorkspaceRoleAssignments` at creation:

```php
$this->workspaceService->createSharedWorkspace(
    $contentRepositoryId,
    $workspaceName,
    $title,
    $description,
    $baseWorkspace,
    WorkspaceRoleAssignments::create(
        WorkspaceRoleAssignment::createForUser(
            $currentUser->getId(),
            WorkspaceRole::MANAGER,
        ),
        WorkspaceRoleAssignment::createForGroup(
            'Neos.Neos:AbstractEditor',
            WorkspaceRole::COLLABORATOR,
        )
    )
);
```


Edit i introduced static factories for that istead:

```php
$this->workspaceService->createSharedWorkspace(
    ...,
    WorkspaceRoleAssignments::createForSharedWorkspace(
        $currentUser->getId()
    )
);
```

```php
$this->workspaceService->createRootWorkspace(
    $this->contentRepository->id,
    WorkspaceName::forLive(),
    WorkspaceTitle::fromString('Public live workspace'),
    WorkspaceDescription::empty(),
    WorkspaceRoleAssignments::createForLiveWorkspace()
);
```


```php

/**
 * Default role assignment to be specified at creation via {@see WorkspaceService::createRootWorkspace()}
 *
 * Users with the role "Neos.Neos:LivePublisher" are collaborators and everybody can read.
 */
public static function createForLiveWorkspace(): self
{
    return new self(
        WorkspaceRoleAssignment::createForGroup(
            'Neos.Neos:LivePublisher',
            WorkspaceRole::COLLABORATOR
        ),
        WorkspaceRoleAssignment::createForGroup(
            'Neos.Flow:Everybody',
            WorkspaceRole::VIEWER
        )
    );
}

/**
 * Default role assignment to be specified at creation via {@see WorkspaceService::createSharedWorkspace()}
 *
 * Users with the role "Neos.Neos:AbstractEditor" are collaborators and the specified user is manager
 */
public static function createForSharedWorkspace(UserId $userId): self
{
    return new self(
        WorkspaceRoleAssignment::createForUser(
            $userId,
            WorkspaceRole::MANAGER,
        ),
        WorkspaceRoleAssignment::createForGroup(
            'Neos.Neos:AbstractEditor',
            WorkspaceRole::COLLABORATOR,
        )
    );
}
```

mhsdesign committed:
TASK: Document how Neos.Neos:LivePublisher and Neos.Neos:AbstractEditor are enforced via the cr core

mhsdesign committed:
BUGFIX: Fix root workspace creation to specify assignments

and replace `createLiveWorkspaceIfMissing`

The `WorkspaceRoleAssignments` have two static factories for sane defaults for the live workspace and share one so that information is not spread out too much over the packages

mhsdesign committed:
TASK: Test current state of ordering if multiple workspaces exist for a user

the order by was just introduced to make this case explicit, the new test pass without.

mhsdesign committed:
TASK: Add test to ensure personal workspace names are always unique

see bug https://github.com/neos/neos-development-collection/issues/2850

mhsdesign committed:
TASK: Document and cleanup `createRootWorkspace` and `createSharedWorkspace`

mhsdesign commented on Neos.Neos/Tests/Behavior/Features/ContentRepository/WorkspaceService.feature:
```diff
@@ -87,33 +87,71 @@ Feature: Neos WorkspaceService related features
+    When a personal workspace for user "test.user" is created
+    Then the personal workspace for user "test.user" is "test-user-1"
+
+  Scenario: For multiple personal workspaces only one workspace is returned
+    When the root workspace "some-root-workspace" is created
+    And the personal workspace "b-user-workspace" is created with the target workspace "some-root-workspace" for user "jane.doe"
+    And the personal workspace "a-user-workspace" is created with the target workspace "some-root-workspace" for user "jane.doe"
+    And the personal workspace "c-user-workspace" is created with the target workspace "some-root-workspace" for user "jane.doe"
+    Then the personal workspace for user "jane.doe" is "a-user-workspace"
```
maybe for the 9.0 release we should disallow having two personal workspaces per user? Not really via sql index but at least via php api? That would allow us to reduce some insanity like seen here 😅 We can always relax that constraint later.

mhsdesign replied:
Edit: with https://github.com/neos/neos-development-collection/pull/5334/commits/31473fb4efae71bb62a10665414a6cd545ae4f4b i introduced a constraint to not allow that 😅

mhsdesign committed:
BUGFIX: Ensure for now that the api does not allow two personal workspaces for one user

we might change the concepts later and add a mapping to the "current" users workspace in their session but for now we dont need the complexity

dlubitz commented on Neos.ContentRepository.Core/Classes/SharedModel/Exception/WorkspaceDoesNotExist.php:
```diff
@@ -14,18 +14,28 @@
+ * @api because exception is thrown during invariant checks on command execution or when attempting to query a non-existing workspace
  */
-final class WorkspaceDoesNotExist extends \DomainException
+final class WorkspaceDoesNotExist extends \RuntimeException
```
Why do we need to change this? All exceptions in this NS are domain exceptions.

Maybe we should rather think about to not usage these domain exceptions outside of this package.

mhsdesign replied:
extracted to https://github.com/neos/neos-development-collection/pull/5399

and i deliberately want to use the `WorkspaceDoesNotExist`, as the api is not forgiving and throws a lot. And its imo a difference if `WorkspaceDoesNotExist` is thrown - which can legally happen - vs when the database is down or something else. If i want to catch something and make a condition out of it it should be a specific exception:


```php
try {
    $workspaceMetadata = $this->workspaceService->getWorkspaceMetadata($contentRepositoryId, $workspaceName);
} catch (WorkspaceDoesNotExist $workspaceDoesNotExist) {
    // something
}
```

dlubitz commented on Neos.Neos/Classes/Domain/Model/WorkspaceRoleAssignments.php:
```diff
@@ -47,4 +96,19 @@ public function count(): int
+        return false;
+    }
+
+    public function add(WorkspaceRoleAssignment $assignment): self
```
`withWorkspaceRoleAssignment` would better indicate the immutability here than `add`.

dlubitz commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -95,25 +97,33 @@ public function setWorkspaceDescription(ContentRepositoryId $contentRepositoryId
         $workspaceName = $this->metadataAndRoleRepository->findPrimaryWorkspaceNameForUser($contentRepositoryId, $userId);
         if ($workspaceName === null) {
-            throw new \RuntimeException(sprintf('No workspace is assigned to the user with id "%s")', $userId->value), 1718293801);
+            throw new \RuntimeException(sprintf('No workspace is assigned to the user with id "%s")', $userId->value), 1733755300);
```
What was wrong with the old code?

dlubitz commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -122,44 +132,64 @@ public function createRootWorkspace(ContentRepositoryId $contentRepositoryId, Wo
-        if ($liveWorkspace !== null) {
-            // live workspace already exists
-            return;
+        $this->metadataAndRoleRepository->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, WorkspaceClassification::ROOT, ownerUserId: null);
```
Named argument is not needed here.
```suggestion
        $this->metadataAndRoleRepository->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, WorkspaceClassification::ROOT, null);
```

mhsdesign replied:
but i like it and its more precise what we are doing here ^^ its internal either way:)

dlubitz replied:
And I still don't like it. It falsely implies that you skipped a parameter - even if it is internal 😝 . I'm not blocking this, but maybe something to discuss in general (code style wise).

bwaidelich replied:
Ah, that was the part you mentioned yesterday :)
My personal take (and I agree, we should update our CGL to the modern language features at some point): I totally agree to the change because it clarifies the `null` (while the other parameters were self-explanatory already).
I tend to use named arguments for these cases more and more – not to skip optional arguments (and even if so: for the reader it wouldn't really matter) but to add context

dlubitz commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -122,44 +132,64 @@ public function createRootWorkspace(ContentRepositoryId $contentRepositoryId, Wo
+                ContentStreamId::create()
+            )
+        );
+        $this->metadataAndRoleRepository->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, WorkspaceClassification::SHARED, ownerUserId: null);
```
```suggestion
        $this->metadataAndRoleRepository->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, WorkspaceClassification::SHARED, null);
```

mhsdesign committed:
TASK: Apply suggestions from review

dlubitz commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/WorkspaceServiceTrait.php:
```diff
@@ -90,17 +117,58 @@ public function aPersonalWorkspaceForUserIsCreated(string $username): void
+        } catch (\Throwable $e) {
+            // todo throw WorkspaceDoesNotExist instead??
+            Assert::assertInstanceOf(\RuntimeException::class, $e, $e->getMessage());
+            Assert::assertSame(1733755300, $e->getCode());
```
```suggestion
            Assert::assertSame(1718293801, $e->getCode());
```

dlubitz reviewed (approved):
Looks good to me, once the tests have been fixed.

dlubitz committed:
BUGFIX: Use correct error code to assert

bwaidelich commented on Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php:
```diff
@@ -716,7 +716,7 @@ public function migrateWorkspaceMetadataToWorkspaceService(\Closure $outputFn):
                 $roleAssignments[] = [
                     'subject_type' => WorkspaceRoleSubjectType::GROUP->value,
-                    'subject' => 'Neos.Neos:Everybody',
+                    'subject' => 'Neos.Flow:Everybody',
```
oops o.O

bwaidelich commented on Neos.Neos/Classes/Domain/Model/WorkspaceRoleAssignments.php:
```diff
@@ -33,6 +44,44 @@ public static function fromArray(array $assignments): self
+     *
+     * Users with the role "Neos.Neos:LivePublisher" are collaborators and everybody can read.
+     */
+    public static function createForLiveWorkspace(): self
+    {
+        return new self(
+            WorkspaceRoleAssignment::createForGroup(
+                'Neos.Neos:LivePublisher',
+                WorkspaceRole::COLLABORATOR
+            ),
+            WorkspaceRoleAssignment::createForGroup(
+                'Neos.Flow:Everybody',
+                WorkspaceRole::VIEWER
+            )
+        );
+    }
```
The original idea was to keep those hard-coded role-assignments out of the core leaving it just to the higher levels (Setup, CLI, Workspace module) to define defaults.
Maybe it's more pragmatic to move that to the core though ¯\_(ツ)_/¯

The many magic strings make me a bit nervous, maybe we can introduce an enum or constant for the "special" roles?

mhsdesign replied:
Actually i tried to improve the current situation as you had `Neos.Neos:LivePublisher` hardcoded _inside_ the service before, now its at least kindof optional if you will, basically a set of sane presets to choose from.
My reasoning was here especially that as the roles are defined inside `Neos.Neos` they ought to be used by this package, otherwise the WorkspaceUi package for example would have the usage to `AbstractEditor` which is odd.

> The many magic strings make me a bit nervous

I like it dangerous :D But yes that how we do it, i mean its originally configuration - unless roles are configured via php we always have that split.

bwaidelich replied:
> Actually i tried to improve the current situation

Yes, it's definitely better than before. And AFAIS it would still allow for "esotheric" scenarios like protected live workspaces.

>> The many magic strings make me a bit nervous
> [...] that how we do it, i mean its originally configuration [...]

![That's how we do it here](https://media.tenor.com/3eoWNFQn-fsAAAAj/that%27s-how-we-do-it-here-devin-montes.gif)

;)

Yes, it originates in configuration. But IMO that's no reason not to introduce constants/enums for the cases we hard-code in PHP (like we already do with the [Neos Administrator role](https://github.com/neos/neos-development-collection/blob/3f65a7435fa24813b8156cd9308c7083089314ed/Neos.Neos/Classes/Security/Authorization/ContentRepositoryAuthorizationService.php#L36))

Anyways, no blocker ofc

bwaidelich replied:
As discussed: I'll take care of some follow-up PR that introduces an enum/constants for the "special" roles and replaces the magic strings here and in other places within Neos

bwaidelich commented on Neos.Neos/Classes/Domain/Repository/WorkspaceMetadataAndRoleRepository.php:
```diff
@@ -319,6 +353,8 @@ public function findPrimaryWorkspaceNameForUser(ContentRepositoryId $contentRepo
                 AND owner_user_id = :userId
         SQL;
+        // We don't order the results and return the first matching workspace.
+        // In case multiple exist - which is not desired currently - and in the happy path not possible via the api (race conditions aside) - the order is not defined.
```
just wondering: couldn't this enforced on the db level?

mhsdesign replied:
actually yes good point i thought at first no, but an unique index on `owner_user_id` AND `content_repository_id` would be possible and makes this super easy. And if we dont need that later we could drop that index.
Will do that next commit!

mhsdesign replied:
Implemented via [TASK: Ensure that the workspace owner is unique via sql index](https://github.com/neos/neos-development-collection/pull/5334/commits/ddccba16f5168bd6c97b439ee729df253c68330a)

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -95,9 +97,7 @@ public function setWorkspaceDescription(ContentRepositoryId $contentRepositoryId
-     * Retrieve the currently active personal workspace for the specified $userId
-     *
-     * NOTE: Currently there can only ever be a single personal workspace per user. But this API already prepares support for multiple personal workspaces per user
+     * Retrieve the personal workspace for the specified user, if no workspace exist an exception is thrown.
```
out of curiosity: why did you remove part of the doc comment?

mhsdesign replied:
because i removed the possibility to have multiple personal workspaces (see https://github.com/neos/neos-development-collection/pull/5334#discussion_r1876050738) any future may happen ideas dont have to be commented in code.

bwaidelich replied:
yep, makes sense!

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -122,44 +132,64 @@ public function createRootWorkspace(ContentRepositoryId $contentRepositoryId, Wo
-        $this->createWorkspace($contentRepositoryId, $workspaceName, $title, $description, $baseWorkspaceName, $ownerId, WorkspaceClassification::PERSONAL);
+        $existingUserWorkspace = $this->metadataAndRoleRepository->findPrimaryWorkspaceNameForUser($contentRepositoryId, $ownerId);
+        if ($existingUserWorkspace !== null) {
+            throw new \RuntimeException(sprintf('Failed to create personal workspace "%s" for user with id "%s", because the workspace "%s" is already assigned to the user', $workspaceName->value, $ownerId->value, $existingUserWorkspace->value), 1733754904);
```
isn't that an `InvalidArgumentException` then?

mhsdesign replied:
could be if you like ^^ maybe `LogicException`? we need to find a good definition at some point ...

`assignWorkspaceRole` for example catches a `UniqueConstraintViolationException` and then throws a `RuntimeException` as well:

> 'Failed to assign role for workspace "%s" to subject "%s" (Content Repository "%s"): There is already a role assigned for that user/group, please unassign that first'

one could argue its the same case of violation of constraints. Id say `RuntimeException` is the better pick, but i agree there should be a better one. Not `InvalidArgumentException` but something in between that says constraints did not met.

bwaidelich replied:
If _I_ like? This is really not about me :)
I just commented on it because there seems to be confusion with exception types.

Personally I used `InvalidArgumentException` when the input did not make sense – on the _type_ level but also _logically_. The latter is not correct, I just learned:
![image](https://github.com/user-attachments/assets/a3f44cb2-e759-4c40-9fa7-9c0edda8f94b)

In the end, it does not really matter that much in this case and `RuntimeException` is totally fine

bwaidelich commented on Neos.Neos/Classes/Domain/Service/WorkspaceService.php:
```diff
@@ -122,44 +132,64 @@ public function createRootWorkspace(ContentRepositoryId $contentRepositoryId, Wo
-            // live workspace already exists
-            return;
+        $this->metadataAndRoleRepository->addWorkspaceMetadata($contentRepositoryId, $workspaceName, $title, $description, WorkspaceClassification::ROOT, ownerUserId: null);
+        foreach ($assignments as $assignment) {
+            $this->metadataAndRoleRepository->assignWorkspaceRole($contentRepositoryId, $workspaceName, $assignment);
```
A single, atomic, operation for the assignments would be nice, should we add some `assignWorkspaceRoles()` method to the repository or even allow for `WorkspaceRoleAssignments|WorkspaceRoleAssignment`?

mhsdesign replied:
okay looked a bit into it, multi insert is not abstracted by dbal and i dont want to write native sql here and risk losing support for postgres or something unexpected ^^: https://github.com/doctrine/dbal/pull/682

multiple operations dont seem as clean but to make it truly atomic the metadata should also be added in the same commit. Maybe we could use transactions?

The operation will never be fully atomic as the content repository workspace is already created - but i guess we could optimise the metadata side.

mhsdesign replied:
implemented via transactional closure via [TASK: Make workspace role and metadata creation atomic](https://github.com/neos/neos-development-collection/pull/5334/commits/08d01e133bc554b1ac9139176a2d6f20f9f0fb54)

bwaidelich reviewed (commented):
Great, thanks a lot!
I just left some (partly nitpicky) comments for now, but in general it's a +1 from me

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/workspace-service-adjustments

mhsdesign committed:
TASK: Ensure that the workspace owner is unique via sql index

We still try to catch the error beforehand via php constraint (so that the `CreateWorkspace` is not handled), but in a race condition this will be thrown at last:

> Integrity constraint violation: 1062 Duplicate entry 'default-janedoe' for key 'owner''

mhsdesign committed:
TASK: Make workspace role and metadata creation atomic
